dipy/reconst/dti.py

```diff
@@ -1801,90 +1850,102 @@ def restore_fit_tensor(design_matrix, data, sigma=None, jac=True,
+                residuals = flat_data[vox] - pred_sig
+                if np.any(np.abs(residuals) > 3 * sigma_vox):
+                    # If you still have outliers, refit without those outliers:
+                    non_outlier_idx = np.where(np.abs(residuals) <= 3 * sigma_vox)
+                    clean_design = design_matrix[non_outlier_idx]
+                    clean_sig = flat_data[vox][non_outlier_idx]
+                    if np.iterable(sigma_vox):
+                        this_sigma = sigma_vox[non_outlier_idx]
+                    else:
+                        this_sigma = sigma_vox
+
+                    if jac:
+                        this_param, status = opt.leastsq(_nlls_err_func,
+                                                         start_params,
+                                                         args=(clean_design,
+                                                               clean_sig,
+                                                               'sigma',
+                                                               this_sigma),
+                                                         Dfun=_nlls_jacobian_func)
+                    else:
+                        this_param, status = opt.leastsq(_nlls_err_func,
+                                                         start_params,
+                                                         args=(clean_design,
+                                                               clean_sig,
+                                                               'sigma',
+                                                               this_sigma))
+
+            # Convert diffusion tensor parameters to the evals and the evecs:
             evals, evecs = decompose_tensor(
                 from_lower_triangular(this_param[:6]))
             params[vox, :3] = evals
             params[vox, 3:12] = evecs.ravel()
 
-        # If leastsq failed to converge and produced nans, we'll resort to the
-        # OLS solution in this voxel:
-        except np.linalg.LinAlgError:
-            this_params = start_params
-            evals, evecs = decompose_tensor(
-                from_lower_triangular(this_params[:6]))
-            params[vox, :3] = evals
-            params[vox, 3:12] = evecs.ravel()
+        # If leastsq failed to converge and produced nans, or because too few
+        # non-outlier data to fit parameters, we'll resort to the OLS solution
+        # in this voxel:
+        except (np.linalg.LinAlgError, TypeError) as e:  # NOTE: catches issue #2574, should be applied to nlls_fit_tensor too
+            if fail_is_nan == False:
+                this_params = start_params
+                evals, evecs = decompose_tensor(
+                    from_lower_triangular(this_params[:6]))
+                params[vox, :3] = evals
+                params[vox, 3:12] = evecs.ravel()
+            else:
+                # NOTE: how should we record the success of RESTORE? Could have
```

Comment:
If RESTORE fails because the sigma was off (as in #2574), raising a warning might be appropriate.

Reply:
Warnings are raised for failed fitting, and option `fail_is_nan` should will indicate when RESTORE has failed